File: packages/npm/src/patterns.ts
import type {
  BudgetRunawayConfig,
  CircularReasoningConfig,
  ErrorSpiralConfig,
  PatternConfig,
  RetryStormConfig,
  StuckAgentConfig,
  TargetedFailureConfig,
  TokenVelocityConfig,
} from "./types.js";

/** Builder helpers for pre-packaged pattern configurations. */
export class Patterns {
  /** Detect repeated identical tool calls within a time window. */
  static retryStorm(config: RetryStormConfig = {}): PatternConfig {
    return { type: "retry_storm", config };
  }

  /** Detect an agent that hasn't produced output after too many steps or too long. */
  static stuckAgent(config: StuckAgentConfig = {}): PatternConfig {
    return { type: "stuck_agent", config };
  }

  /** Detect repeated tool call failures within a time window. */
  static errorSpiral(config: ErrorSpiralConfig = {}): PatternConfig {
    return { type: "error_spiral", config };
  }

  /** Detect cumulative LLM cost or token usage exceeding thresholds. */
  static budgetRunaway(config: BudgetRunawayConfig = {}): PatternConfig {
    return { type: "budget_runaway", config };
  }

  /** Detect sudden spikes in token consumption rate per step. */
  static tokenVelocity(config: TokenVelocityConfig = {}): PatternConfig {
    return { type: "token_velocity", config };
  }

  /** Detect repeating cycles in tool call sequences. */
  static circularReasoning(config: CircularReasoningConfig = {}): PatternConfig {
    return { type: "circular_reasoning", config };
  }

  /** Detect repeated failures on the same target (test, file, endpoint) within a session. */
  static targetedFailure(config: TargetedFailureConfig = {}): PatternConfig {
    return { type: "targeted_failure", config };
  }

  /** All default patterns with default configurations. */
  static defaults(): PatternConfig[] {
    return [
      Patterns.retryStorm(),
      Patterns.stuckAgent(),
      Patterns.errorSpiral(),
      Patterns.budgetRunaway(),
      Patterns.tokenVelocity(),
      Patterns.circularReasoning(),
    ];
  }
}

File: packages/python/src/varpulis_agent_runtime/patterns.py
"""Pattern configuration builders."""

from __future__ import annotations

from typing import Any


class Patterns:
    """Builder helpers for pre-packaged pattern configurations."""

    @staticmethod
    def retry_storm(**kwargs: Any) -> dict[str, Any]:
        return {"type": "retry_storm", "config": kwargs}

    @staticmethod
    def stuck_agent(**kwargs: Any) -> dict[str, Any]:
        return {"type": "stuck_agent", "config": kwargs}

    @staticmethod
    def error_spiral(**kwargs: Any) -> dict[str, Any]:
        return {"type": "error_spiral", "config": kwargs}

    @staticmethod
    def budget_runaway(**kwargs: Any) -> dict[str, Any]:
        return {"type": "budget_runaway", "config": kwargs}

    @staticmethod
    def token_velocity(**kwargs: Any) -> dict[str, Any]:
        return {"type": "token_velocity", "config": kwargs}

    @staticmethod
    def circular_reasoning(**kwargs: Any) -> dict[str, Any]:
        return {"type": "circular_reasoning", "config": kwargs}

    @staticmethod
    def defaults() -> list[dict[str, Any]]:
        return [
            Patterns.retry_storm(),
            Patterns.stuck_agent(),
            Patterns.error_spiral(),
            Patterns.budget_runaway(),
            Patterns.token_velocity(),
            Patterns.circular_reasoning(),
        ]
